cdap-app-fabric/src/main/java/io/cdap/cdap/internal/sysapp/SystemAppManagementService.java

```diff
@@ -0,0 +1,135 @@
+  private ExecutorService executorService;
+
+  @Inject
+  SystemAppManagementService(CConfiguration cConf, ApplicationLifecycleService appLifecycleService,
+                             ProgramLifecycleService programLifecycleService) {
+    this.systemAppConfigDirPath = new File(cConf.get(Constants.SYSTEM_APP_CONFIG_DIR));
+    this.systemAppEnableExecutor = new SystemAppEnableExecutor(appLifecycleService,
+        programLifecycleService);
+  }
+
+  @Override
+  protected void startUp() {
+    LOG.debug("Starting {}", getClass().getSimpleName());
+    executorService = Executors
+        .newSingleThreadExecutor(Threads.createDaemonThreadFactory("sys-app-management-service"));
+    executorService.execute(() -> {
+      try {
+        // Execute all steps for each config file in system config directory.
+        bootStrapSystemAppConfigDir();
+      } catch (Exception ex) {
+        LOG.error("Got exception in watch service for system app config dir", ex);
+      }
+    });
+  }
+
+  private void bootStrapSystemAppConfigDir() throws Exception {
+    LOG.debug("Number of config files {} in system app config directory.",
+        systemAppConfigDirPath.listFiles().length);
+
+    for (File sysAppConfigFile : systemAppConfigDirPath.listFiles()) {
+      LOG.debug("Running steps in config file {}", sysAppConfigFile.getAbsoluteFile());
+      executeSysConfig(sysAppConfigFile.getAbsoluteFile());
+    }
+  }
+
+  private void executeSysConfig(File fileName) throws Exception {
+    SystemAppConfig config = parseConfig(fileName);
+    for (SystemAppStep step : config.getSteps()) {
+      try {
+        step.validate();
+      } catch (IllegalArgumentException e) {
+        LOG.warn("Config step {} failed because it is malformed: {}", step.getLabel(),
+            e.getMessage());
+        return;
+      }
+      if (step.getType() == SystemAppStep.Type.ENABLE_SYSTEM_APP) {
+        systemAppEnableExecutor.deployAppAndStartPrograms(step.getArguments());
+        LOG.debug("Deployed and enabled system app with id {} and label {}. Config file: {}.",
+            step.getArguments().getId(), step.getLabel(), fileName);
+      } else {
+        LOG.warn("Unknown system app step type {} for step label {}. Config file: {}. Ignoring it.",
+            step.getLabel(), step.getType(), fileName);
+      }
+    }
+  }
+
+  private SystemAppConfig parseConfig(File fileName) {
+    try {
+      try (Reader reader = new FileReader(fileName)) {
+        return GSON.fromJson(reader, SystemAppConfig.class);
+      }
+    } catch (FileNotFoundException e) {
+      LOG.info("System app config file {} does not exist.",
+          fileName);
+      return SystemAppConfig.EMPTY;
+    } catch (JsonParseException e) {
+      LOG.warn("Could not parse system app config file {}.",
+          fileName, e);
+      return SystemAppConfig.EMPTY;
+    } catch (IOException e) {
+      LOG.warn("Could not system app config file {}.",
```

Comment:
Could not read system app config file

Reply:
You can combine the `FileNotFoundException` as the `IOException`. Just log with the exception would give enough information to debug.